## Gestire dati mancanti

In [3]:
import pandas as pd
import numpy as np

In [4]:
BASE_URL = "https://raw.githubusercontent.com/ProfAI/machine-learning-fondamenti/refs/heads/main/datasets/"

In [5]:
iris = pd.read_csv(BASE_URL + "iris_missing.csv")
iris.head()

,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,NaN,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,NaN,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa


In [4]:
iris.shape

(150, 5)

In [6]:
iris.count()

sepal_length    147
sepal_width     140
petal_length    148
petal_width     150
species         150
dtype: int64

In [10]:
iris.isna().sum()

sepal_length     3
sepal_width     10
petal_length     2
petal_width      0
species          0
dtype: int64

Metodo 1: Rimuovere righe o colonne con valori mancanti

In [9]:
iris_drop = iris.copy()

In [24]:
iris_drop = iris.dropna()
iris_drop.head()

,sepal_length,sepal_width,petal_length,petal_width,species
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa
6,4.6,3.4,1.4,0.3,setosa
7,5.0,3.4,1.5,0.2,setosa


In [15]:
iris_drop.shape

(135, 5)

In [16]:
iris.shape

(150, 5)

In [27]:
iris_drop = iris_drop.dropna(axis=1)
iris_drop.head()

,petal_width,species
0,0.2,setosa
1,0.2,setosa
2,0.2,setosa
3,0.2,setosa
4,0.2,setosa


In [28]:
iris_drop.shape

(150, 2)

In [6]:
tresh = int(iris.shape[0]*0.9)
tresh

135

<b> thresh </b> serve a stabilire una soglia minima di valori validi che una riga o una colonna deve avere per essere mantenuta.

Quindi:

- se lo usi sulle righe, tieni solo le righe che hanno almeno un certo numero di valori non mancanti
- se lo usi sulle colonne, tieni solo le colonne che hanno almeno un certo numero di valori non mancanti

tresh = int(iris.shape[0] * 0.9) serve a trasformare una percentuale in un numero assoluto di righe valide richieste.
- iris.shape[0] restituisce il numero totale di righe del dataset, che nell'esempio sono 150
- 150 * 0.9 = 135
- int(...) converte il risultato in intero, quindi 135

Quindi tresh = 135 significa:

voglio tenere solo le colonne che hanno almeno 135 valori non mancanti
oppure, detto in modo equivalente, voglio scartare le colonne che hanno troppi valori mancanti

Si fa questa operazione perché il parametro thresh di dropna() *non accetta una percentuale*, ma un numero minimo di valori validi.
Se tu vuoi dire “tieni solo le colonne con almeno il 90% di dati presenti”, devi prima convertire quel 90% in un numero, cioè 135 nel tuo caso.

In [8]:
iris_drop = iris_drop.dropna(axis=1, thresh=145) #treshold = specifichiamo i valori minimi validi che devono essere in una colonna
iris_drop.head()

NameError: name 'iris_drop' is not defined

nel caso delle righe, basta non definire l'asse

In [31]:
iris_drop.shape

(150, 4)

se vogliamo rimuovere tutte le colonne e le righe che hanno valori mancanti:

In [ ]:
iris_drop = iris_drop.dropna(how="all")

se vogliamo solo determinate colonne:

In [11]:
iris_drop = iris_drop.dropna(subset= ["petal_width"])
iris_drop.head(10)

,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,NaN,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,NaN,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa
5,NaN,3.9,1.7,0.4,setosa
6,4.6,3.4,1.4,0.3,setosa
7,5.0,3.4,1.5,0.2,setosa
8,4.4,2.9,1.4,0.2,setosa
9,4.9,3.1,1.5,0.1,setosa


in questo caso verranno rimosse le righe dove il valore della colonna selezionata non è disponibile.